Add gen 7 starter flair setting. Fixes #549

YaManicKill opened the pull request.

Raia commented:
Not sure about SVeX for this, as we have the ribbon occupying the second slot. I guess it doesn't hurt to add the class if it doesn't do anything there?

YaManicKill commented:
Alright. I assumed you were doing it on both (as we have upgrade and hok on
trades) but I can remove it from svex.

On Sat, 12 Nov 2016, 17:25 Raia, <notifications@github.com> wrote:

> Not sure about SVeX for this, as we have the ribbon occupying the second
> slot. I guess it doesn't hurt to add the class if it doesn't do anything
> there?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/pokemontrades/flairhq/pull/552#issuecomment-260135479>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACD_4QxcK3T3D4P3oohqeRgcKvOQxMyGks5q9fZ2gaJpZM4KwctL>
> .
>

Raia commented:
Nobody actually has hok on trades anymore. I think the only thing that would conflict on trades is `upgrade` and that only affects ~5 inactive people :p

YaManicKill commented:
Aight. Just trades, then.

On Sat, 12 Nov 2016, 17:30 Raia, <notifications@github.com> wrote:

> Nobody actually has hok on trades anymore. I think the only thing that
> would conflict on trades is upgrade and that only affects ~5 inactive
> people :p
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/pokemontrades/flairhq/pull/552#issuecomment-260135774>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACD_4Q01YNJIS-QpZYb3xlGGHHB5Omfpks5q9feLgaJpZM4KwctL>
> .
>

YaManicKill commented:
Alrighty, this should be sorted, now. I have removed svex from the setting, and I have updated the test code to override the local.js config file (for the server, and it also means we don't have to change the tests when this is removed from live).

not-an-aardvark commented on test/.eslintrc:
```diff
@@ -1,6 +1,7 @@
   }
-}
\ No newline at end of file
+}
```
We can replace this entire file with:

```json
{"env": "mocha"}
```

not-an-aardvark commented on assets/styles/flairs.less:
```diff
@@ -97,3 +97,23 @@ a[href="/lucky"], a[href="/egg"], a[href="/eevee"], a[href="/togepi"], a[href="/
 .flair2-smartribbon:before { background-position: -66px 0; margin-bottom: -2px; }
 .flair2-toughribbon:before { background-position: -88px 0; margin-bottom: -2px; }
+
+.flair-rowlet:after, .flair-litten:after, .flair-popplio:after {
```
If we're hardcoding the extra flairs into the stylesheet, I think we might as well hardcode them into the rest of the codebase as well, rather than putting an array in the config file.

YaManicKill replied:
Keeping the extra flairs in the config makes it easier to turn off the setting (I realise I haven't added a setting to remove it from the frontend, but meh). A **bit** of config is better than 0 config.

not-an-aardvark replied:
Fair enough

not-an-aardvark commented on api/services/Flairs.js:
```diff
@@ -3,6 +3,8 @@ var _ = require('lodash');
 var referenceService = require('./References.js');
 var NodeCache = require('node-cache');
 var app_claim_cache = new NodeCache({stdTTL: 300});
+var extraFlair = require('../../config/local.js').extraFlair;
```
I would recommend using `sails.config.extraFlair` instead, so that it can be overridden in other config files if necessary.

YaManicKill replied:
Yeah, I guess so.

not-an-aardvark commented on api/controllers/FlairController.js:
```diff
@@ -105,13 +105,17 @@ module.exports = {
       newsvFlair = newsvFlair.replace(/2/, "");
       var promises = [];
+      var extraFlair = req.allParams().extraFlair;
+      if (extraFlair) {
```
We should validate that the extra flair matches one of the legal flairs in the config file, otherwise people can put arbitrary text in their flair.

not-an-aardvark commented on api/services/Flairs.js:
```diff
@@ -261,6 +263,13 @@ exports.makeNewCSSClass = function (previous_flair, new_addition, subreddit) {
   if (new_addition === 'involvement') {
     return previous_flair.replace(/( |$)/, '1$1');
   }
+  if (extraFlair.includes(new_addition)) {
```
If we're using `Array#includes`, we can no longer support Node 4. I'm fine with this, but we might want to update the readme.

YaManicKill replied:
Yeah, I just realised that when the build failed again.

YaManicKill replied:
What are we running on the server just now? 6?

not-an-aardvark replied:
I think we're running 4.6.2 at the moment. We can probably upgrade to 6 soon.

YaManicKill replied:
Alright, well if we don't want to upgrade to 6 at the same time as this release, I'll need to move this back. I assumed the server was using babel, is it not?

not-an-aardvark replied:
I don't remember. I think we're still using Babel 5, and I forget how that works with regard to polyfills.

not-an-aardvark commented on api/services/Flairs.js:
```diff
@@ -261,6 +263,13 @@ exports.makeNewCSSClass = function (previous_flair, new_addition, subreddit) {
   }
+  if (extraFlair.includes(new_addition)) {
+    if (previous_flair.match(extraFlairRegExp)) {
+      return previous_flair.replace(extraFlairRegExp, new_addition);
```
This will fail if someone's IGN is `litten`.

not-an-aardvark commented on api/services/Flairs.js:
```diff
@@ -261,6 +263,13 @@ exports.makeNewCSSClass = function (previous_flair, new_addition, subreddit) {
+    if (previous_flair.match(extraFlairRegExp)) {
+      return previous_flair.replace(extraFlairRegExp, new_addition);
+    } else {
+      return previous_flair += (" " + new_addition);
```
Is this supposed to just be `+` rather than `+=`?

YaManicKill replied:
Meh, i guess so.
